### ISSUE TITLE
JDBC Java - Memory Leak - Linux in Docker

### ISSUE THREAD

**Issue by brianwyka** [duckdb/duckdb#9712]
### What happens?

DuckDB is leaking memory even after closing connections and statements.  We have a multi-threaded architecture which is opening connections to the in-memory DuckDB and creating unique tables, ingestion data into them, writing them to file, and then dropping the tables after complete.  Each thread creates a new connection and closes it when complete.  No 2 threads ever act upon the same table.

Please see the attached image for visual of the docker container resources.

Perhaps somewhat related or linked to #9263 or #109?  Maybe a leak in `httpfs` or nested `s3fs` extension layer?

<img width="1021" alt="image" src="https://github.com/duckdb/duckdb/assets/46069296/b7084288-299c-4f59-93c4-970162d087dd">

### To Reproduce
Launch the following code in parallel threads (I was executing 5-10 concurrently in my workload).

```java
try (var connection = DriverManager.getConnection("jdbc:duckdb:")) {

    // Used for unique table name
    final String ingestionId = UUID.randomUUID().toString().replace("-", "");

    // Setup connection to S3
    try (var statement = connection.createStatement()) {
        final String setupSql = """
            SET s3_region="us-east-1";
            SET s3_access_key_id="***";
            SET s3_secret_access_key="***";
            SET s3_session_token="***";
        """;
        statement.executeUpdate(setupSql);
    }

    // Create temporary table
    try (var statement = connection.createStatement()) {
        final String createTableSql = """
            CREATE TABLE table_%s (
                id VARCHAR,
                name VARCHAR, 
                description VARCAR
            );
        """.formatted(ingestionId);
        statement.executeUpdate(createTableSql);
    }

   // Read GZIP compressed CSV data from S3 into temporary table
   try (var statement = connection.createStatement()) {
        final String readSql = """
            INSERT INTO table_%s
            SELECT * FROM read_csv_auto('s3://bucket/file.csv.gz', ALL_VARCHAR=true);
        """.formatted(ingestionId);
        statement.executeUpdate(readSql);
    }

    // Export the data from DuckDB to parquet files on file system
    try (var statement = connection.createStatement()) {
        final String writeSql = """
            COPY (
                SELECT *
                FROM table_%s 
                ORDER BY 
                  column1 ASC
            )
            TO '/tmp/%s.parquet'
            WITH (FORMAT PARQUET, COMPRESSION GZIP);
        """.formatted(ingestionId, ingestionId);
        statement.executeUpdate(writeSql);
    }

    // Drop the temporary table to cleanup memory (or at least try to)
    try (var statement = connection.createStatement()) {
        final String dropSql = "DROP TABLE table_%s;".formatted(ingestionId);
        statement.executeUpdate(dropSql);
    }

}
```

### OS:

**Host Machine:**
Mac iOS x64
`Darwin Version 22.4.0: Mon Mar  6 21:00:17 PST 2023; root:xnu-8796.101.5~3/RELEASE_X86_64 x86_64`

**Docker Container:**
`Linux 6.4.16-linuxkit #1 SMP PREEMPT_DYNAMIC Fri Nov 10 14:51:57 UTC 2023 x86_64 GNU/Linux`

### DuckDB Version:

`0.9.2`

### DuckDB Client:

Java JDBC

### Full Name:

Brian Wyka

### Affiliation:

LogicMonitor

### Have you tried this on the latest `main` branch?

I have tested with a release build (and could not test with a main build)

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

**Comment by Mause** [comment]
Can you please clarify your OS? iOS on an iPhone or MacOS on a Mac?

**Comment by brianwyka** [comment]
Updated @Mause

**Comment by brianwyka** [comment]
I did some additional troubleshooting to log some diagnostic information and here is what I get after all work with the DB is complete:

```
PRAGMA database_size
```

```
[DuckDbManager] DuckDB state - databaseSize=0 bytes, blockSize=0, totalBlocks=0, usedBlocks=0, freeBlocks=0, walSize=0 bytes, memoryUsage=0 bytes, memoryLimit=1.5GB
```

So, it thinks its not using any memory any more, but has it cached things in memory that is not reported here?

**Comment by brianwyka** [comment]
I just ran another iteration of this where I read from S3 directly to file system parquet file without an intermediate table in DuckDB and it still results in the same memory issue, not reclaimed.

**Comment by brianwyka** [comment]
Some more insight.  Tested this outside of a Docker container and see that the JVM heap is reclaimed as expected.  So seems the memory leak is native memory on the container?
<img width="965" alt="image" src="https://github.com/duckdb/duckdb/assets/46069296/12769ed6-f42a-41bd-90c7-a4abd49f5be8">

**Comment by brianwyka** [comment]
I just ran another experiment where I removed the `INSERT INTO...` and still have the memory problem.

```java
   // Read GZIP compressed CSV data from S3 into temporary table
   try (var statement = connection.createStatement()) {
        final String readSql = """
            SELECT * FROM read_csv_auto('s3://bucket/file.csv.gz', ALL_VARCHAR=true);
        """.formatted(ingestionId);
        statement.executeUpdate(readSql);
    }
```

So, based on this, seems likely that the leak is in the `SELECT` in `httpfs` / `s3fs` layer...

Here is more

**Comment by brianwyka** [comment]
Got some more information from a heap dump.  Seeing lots of `HeapByteBuffer` references remaining after completion of work, after closing out all connections and statements and dropping tables.  I did see that The JDBC java library is using `ByteBuffer` to interop with the DuckDB native binary, maybe this could be helpful...

<img width="1307" alt="image" src="https://github.com/duckdb/duckdb/assets/46069296/b8857f7a-a280-4111-a270-bc3b27709034">

**Comment by brianwyka** [comment]
Did some more testing, and here is some more information.

Introducing more load reclaims memory after each execution, but never returns the original memory which was claimed.
<img width="954" alt="image" src="https://github.com/duckdb/duckdb/assets/46069296/0d7036fe-f367-4fdc-9cc6-d998b25ff79f">

I took a heap dump of the application, and it does not add up to the container memory.
<img width="954" alt="image" src="https://github.com/duckdb/duckdb/assets/46069296/a63910ae-f12c-40e4-b91c-c390640232c5">

So the memory in the container is most likely a result of leak in JNI layer.

We are using AWS CRT for S3 async client outside of DuckDB usage, which also uses JNI, but we are closing our resources and that has been confirmed.

**Comment by brianwyka** [comment]
It seems that the higher concurrency, the worse the problem, although the memory leak is there even without concurrency.  I've also noticed that the `memory_limit` option is not really respected at all, as far as container memory usage is concerned...

Next, I'll try to get `vagrind` in the container to investigate further.

**Comment by brianwyka** [comment]
@Mause , here is a self-contained reproducer:

## Dockerfile:
```dockerfile
FROM amazoncorretto:17.0.9

WORKDIR /tmp

COPY /path/to/jar.jar app.jar

# JMX
EXPOSE 9003

# Run JAR with JMX arguments
CMD [ \
  "java", \
  "-Djava.rmi.server.hostname=0.0.0.0", \
  "-Dcom.sun.management.jmxremote", \
  "-Dcom.sun.management.jmxremote.port=9003", \
  "-Dcom.sun.management.jmxremote.rmi.port=9003", \
  "-Dcom.sun.management.jmxremote.ssl=false", \
  "-Dcom.sun.management.jmxremote.authenticate=false", \
  "-Djava.rmi.server.hostname=0.0.0.0", \
  "-jar", \
  "/tmp/app.jar" \
]
```

## Java Code
```java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.sql.Statement;
import java.time.Duration;
import java.util.List;
import java.util.Properties;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.stream.IntStream;

public class MemoryLeak {

    private static final String SETUP_SQL = """
        INSTALL httpfs;
        LOAD httpfs;
        SET session s3_region='%s';
        SET session s3_access_key_id='%s';
        SET session s3_secret_access_key='%s';
        SET session s3_session_token='%s';
    """;

    private static final String SELECT_SQL = """
        CREATE TABLE temp_%d AS 
        SELECT *
        FROM read_csv_auto('%s', ALL_VARCHAR=true);
    """;

    private static final String DROP_SQL = """
        DROP TABLE IF EXISTS temp_%d;
    """;

    private static final String MEMORY_LIMIT = "5GB";
    private static final int TASK_COUNT = 10;
    private static final int CONCURRENCY = 2;

    /**
     * A CSV file in S3 which is compressed with GZIP
     */
    private static final String S3_KEY = "s3://bucket/key.csv.gz"; // TODO: Replace with an S3 URI to compressed VSC file (~100MB in size)


    public static void main(final String[] args) throws InterruptedException, SQLException {

        // Thread pool setup
        final ExecutorService threadPool = Executors.newFixedThreadPool(CONCURRENCY);

        // Create connection properties
        final Properties connectionProperties = new Properties();
        connectionProperties.setProperty("access_mode", "READ_WRITE");

        try (final Connection connection = DriverManager.getConnection("jdbc:duckdb:", connectionProperties);
             final Statement statement = connection.createStatement()) {
            statement.execute("SET memory_limit='" + MEMORY_LIMIT + "'");
        }

        // Sleep for a while to let diagnostic tools warmup
        System.out.println("Sleeping for 30 seconds to let container diagnostic tools warmup");
        Thread.sleep(Duration.ofSeconds(30).toMillis());

        System.out.println("Beginning tasks (" + TASK_COUNT + ") with concurrency (" + CONCURRENCY + ")");

        // Create workloads on the executor
        final List<CompletableFuture<Void>> completableFutures = IntStream.range(1, TASK_COUNT + 1) // tasks
            .mapToObj(taskNumber -> CompletableFuture.runAsync(() -> quack(connectionProperties, taskNumber), threadPool))
            .toList();

        // Wait for all to complete
        final CompletableFuture.allOf(completableFutures.toArray(CompletableFuture[]::new))
            .whenComplete((v, e) -> threadPool.shutdown())
            .join();

        System.out.println("Sleeping for 10 minutes to keep container running to allow diagnostic analysis");

        // Sleep for a while to keep app running to monitor memory after workload completes
        Thread.sleep(Duration.ofMinutes(10).toMillis());

    }

    private static void quack(final Properties connectionProperties, final int taskNumber) {
        System.out.println("[" + taskNumber + "] > Opening Connection");
        try (final Connection connection = DriverManager.getConnection("jdbc:duckdb:", connectionProperties)) {

            // Setup the s3 variables
            final String s3Region = "us-east-1"; // TODO: replace with region of bucket
            final String s3AccessKeyId = ""; // TODO: replace with access key id
            final String s3SecretAccessKey = ""; // TODO: replace with secret access key
            final String s3SessionToken = ""; // TODO: replace with session token
            try (final Statement statement = connection.createStatement()) {

                System.out.println("[" + taskNumber + "] > Running Setup SQL");
                statement.execute(SETUP_SQL.formatted(s3Region, s3AccessKeyId, s3SecretAccessKey, s3SessionToken));

                System.out.println("[" + taskNumber + "] > Running SELECT SQL");
                statement.execute(SELECT_SQL.formatted(taskNumber, S3_KEY));

                System.out.println("[" + taskNumber + "] > Running DROP SQL");
                statement.execute(DROP_SQL.formatted(taskNumber));
            }

        } catch (final SQLException e) {
            System.err.println("[" + taskNumber + "] > Error: " + e.getMessage());
        } // connection and statement closed automatically with "try-with-resources"

        System.out.println("[" + taskNumber + "] > Closed Connection");
    }
```

**Comment by mvrueden** [comment]
Regarding the max_memory property or memory_limit it is not considered, even when running outside docker. I have seen it. However if limit is set too low I get out of memory exception. Maybe this is related somehow?

**Comment by Mause** [comment]
The memory limit in this example code will never be considered no, as it's set on a database that is immediately destroyed (add it to the connection proprieties if you want it to be actually used)

**Comment by brianwyka** [comment]
Thanks @Mause , I didn't see this in the documentation anywhere, going to give this a try to see if there is a change in behavior.

**Comment by brianwyka** [comment]
Here are the memory usage results inside linux docker container after the `memory_limit` being set on the connection properties:

```java
    private static Properties getConnectionProperties() {
        final Properties connectionProperties = new Properties();
        connectionProperties.setProperty("access_mode", "READ_WRITE");
        connectionProperties.setProperty("memory_limit", "8GB");
        return connectionProperties;
    }
```

<img width="441" alt="Screenshot 2023-12-11 at 09 37 23" src="https://github.com/duckdb/duckdb/assets/46069296/7d3a55e1-3b31-480e-aa70-23006899753f">

The `8GB` was respected from a total container perspective up until the very end, when the leak finally pushed it over the edge.  The memory consumption was much more predictable with this as a connection property but did not make any stride towards mitigating the memory leak, which should be apparent in the graph still...

**Comment by brianwyka** [comment]
I do see that after some time of no activity, all connections have been closed and released, that some memory slowly gets reclaimed:
<img width="437" alt="image" src="https://github.com/duckdb/duckdb/assets/46069296/2f094735-bd07-4165-9013-ec0519facf50">

**Comment by viviakemik** [comment]
hello! 

Please, do you have any news on this issue? I tested in two different environments in **java 8 + duckdb 0.10.0 + windows** and in **java 8 + duckdb 0.10.0 + linux on kubernetes** (on windows it doesn't seem to have the memory leak problem, but on kubernetes it does). I also tried using jelastic with` -javaagent:/usr/apache-tomcat/shared/lib/jelastic-gc-agent.jar=period=60` but it had no effect on memory. After 2 executions the memory drops but not everything. However, after many executions in a row, the memory remains always high. I'm using duckdb in memory and I use appender to then write the parquet files and do some queries in these files.

![image](https://github.com/duckdb/duckdb/assets/52158678/53586e03-d55e-4d58-b0a8-f06dd54f72ca)


Thanks!

**Comment by mvrueden** [comment]
Not sure if related to your issues here, but I had high memory consumption which I misread as a memory leak. I was able to fix this by switching to streaming mode. With that option also memory settings were honored. I only consumed parquet files. Maybe this is helpful information.

**Comment by brianwyka** [comment]
> > hello!
> > Please, do you have any news on this issue? I tested in two different environments in **java 8 + duckdb 0.10.0 + windows** and in **java 8 + duckdb 0.10.0 + linux on kubernetes** (on windows it doesn't seem to have the memory leak problem, but on kubernetes it does). I also tried using jelastic with` -javaagent:/usr/apache-tomcat/shared/lib/jelastic-gc-agent.jar=period=60` but it had no effect on memory. After 2 executions the memory drops but not everything. However, after many executions in a row, the memory remains always high. I'm using duckdb in memory and I use appender to then write the parquet files and do some queries in these files.
> > ![image](https://private-user-images.githubusercontent.com/52158678/306879033-53586e03-d55e-4d58-b0a8-f06dd54f72ca.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkxOTg3MjgsIm5iZiI6MTcwOTE5ODQyOCwicGF0aCI6Ii81MjE1ODY3OC8zMDY4NzkwMzMtNTM1ODZlMDMtZDU1ZS00ZDU4LWIwYTgtZjA2ZGQ1NGY3MmNhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI5VDA5MjAyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ4ODA5NzJlMjAxZjI1MWI1NjNlYWM1OWUxZDA5OWI2NmIwZDA3OTYwNTkwYTAyYzZlNWE1YzQyOGQ1ZGFlMGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.y8fnwYsxAbnVlUy20BaViCwEkLPFw1NL6iHHH1PnRRQ)
> > Thanks!
> 
> I found the reason may be relation to jemalloc which is used by linux glibc. You can do malloc_trim(); after close the database object to release the memory, which I used and it takes effect.

@nervmor , if we have concurrent connections to the database, we can't call `malloc_trim()` until after all are closed, correct?  Also, from Java, how would we invoke this command?

**Comment by brianwyka** [comment]
@nervmor , is the "database object you are referring to, the JDBC `Connection` object?  In my case, I am not using the row appender.  I'm doing reads from S3 directly into a parquet file.

I'm just worried if `malloc_trim()` is called by one thread while another is doing work in the in-memory database, that there will be memory corruption...

**Comment by nervmor** [comment]
@brianwyka You should close the database object after all thread which used it.And then call malloc_trim to release the idle  memory. JDBC Connection is the database object.

**Comment by brianwyka** [comment]
So, I'm assuming that this has to be called from the DuckDB C API, and we cannot invoke it from Java.  Does that mean a change is required with DuckDB @nervmor ?

**Comment by nervmor** [comment]
@brianwyka Perhaps you need call malloc_trim by c code

**Comment by nervmor** [comment]
@brianwyka FYI https://android.googlesource.com/platform/external/jemalloc_new/+/6e6a93170475c05ebddbaf3f0df6add65ba19f01/TUNING.md

**Comment by lnkuiper** [comment]
@brianwyka yes. This should fix issues with DuckDB holding on to memory after closing the connection in all of our clients.

I hope this fixes your issue too.

**Comment by lnkuiper** [comment]
@brianwyka, the fix has been added to the latest 0.10.1 release. Has this fixed the issue for you? If yes, please let us know if we can close the issue!

**Comment by brianwyka** [comment]
@lnkuiper , waiting for #11244 to be fixed first.

**Comment by rajmudgil** [comment]
@lnkuiper This issue is still occurring in latest version (0.10.1). Even after closing the connection, memory consumed is not getting freed up.

**Comment by rajmudgil** [comment]
@lnkuiper I'm utilizing a Node.js client and encountering the same problem within a Linux Docker container. As mentioned in the prior comment, I'm employing a specific function to run all DuckDB queries. However, despite closing the connection, the memory consumption does not decrease. You mentioned that in version 0.10.1, memory should be released after closing the connection across all client types, but this doesn't appear to be the case with Node.js.

**Comment by lnkuiper** [comment]
@rajmudgil thanks for the information. The Node client is currently not built with jemalloc, so the implemented fix does not affect Node. I'm unsure whether this will be resolved by enabling jemalloc. If there is still memory being held by the client, that might be a Node-specific issue.

**Comment by brianwyka** [comment]
@lnkuiper, things look promising in some local tests I have done with the release download JAR, but I won't know for sure until I test out on a larger dataset.  Will try that sometime in near future.

**Comment by rajmudgil** [comment]
@lnkuiper You mentioned that the Node client isn't constructed using jemalloc, yet we're encountering a similar issue in our production deployment with the Node client. Despite thorough checks, we haven't detected any problems or memory leaks in our code. Do you have any recommendations? Could using duckdb to release memory after query execution be a viable solution?

**Comment by lnkuiper** [comment]
@lnkuiper Please file a separate issue with a reproducible example for the Node client if you can. Then we can we look into it.

**Comment by brianwyka** [comment]
@lnkuiper , did a test on a larger dataset today and still seeing issues with memory running out.
<img width="468" alt="image" src="https://github.com/duckdb/duckdb/assets/46069296/6002034a-8d01-4647-a47c-e07e2f5f4d02">

Had `10` parallel threads in `JDBC` with `10`CPUs allocated to the container.  Container had `20Gi` overall memory allocated to it within Kubernetes.

**Comment by lnkuiper** [comment]
@brianwyka If you are reading many, many Parquet files, potentially our Parquet metadata cache is accumulating memory. Try `set enable_object_cache=false;`. If that doesn't help, the issue might not be a general DuckDB problem, but a JDBC specific problem.

**Comment by brianwyka** [comment]
@lnkuiper , according to the [configurations page](https://duckdb.org/docs/configuration/overview.html), `false` is already the default value for `enable_object_cache`.

**Comment by lnkuiper** [comment]
@brianwyka, right, then I suspect the issue to be in the JDBC client.

**Comment by brianwyka** [comment]
I don't think its anything to do with the JDBC client @lnkuiper, but that's just my suspicion.  I've run tests outside of linux container with JDBC on the mac and seems to be fine.

**Comment by lnkuiper** [comment]
@brianwyka jemalloc cleans up memory synchronously, i.e., cleanup is triggered by (de-)allocations. If DuckDB is idle, memory is never cleaned up. I've reworked the jemalloc extension, but the rework will only be added when 1.1 is released. The rework tries to clean up memory more quickly, and also adds a setting `SET allocator_background_threads=true` so that a background thread asynchronously performs cleanup.

**Comment by brianwyka** [comment]
@lnkuiper , is a `DROP` statement considered a de-allocation?  Not sure how to shrink the memory footprint otherwise, since that isn't working.

**Comment by lnkuiper** [comment]
@brianwyka, most statements will cause allocation activity, so I think `DROP TABLE` should work. However, if you are using DuckDB tables, and still have the connection open, DuckDB will keep persistent data cached in memory, like most database systems do, to speed up repeated querying of the same data.

If you want to make sure DuckDB evicts all persistent data, you can set a low memory limit, or close the DuckDB connection. Again, the cleanup upon closing should be better with the changes that will come in 1.1.0.

**Comment by brianwyka** [comment]
@lnkuiper , currently, I have one in memory connection which lives forever in the java app, and I create all other connections by calling the `DuckDBConnection.duplicate()` method.  The connections that I get from the duplication are always closed out immediately after use, however, the long living one is not. Does this present an issue?

**Comment by lnkuiper** [comment]
@brianwyka, you can have multiple connections to the same DB, but there will only be one DB instance. The persistent data that is cached in memory is shared across all open connections, so, as long as that one connection stays open, DuckDB will keep some persistent data cached in memory (if there is any).

You can force the persistent data out by setting a low memory limit.

**Comment by lnkuiper** [comment]
@brianwyka that's interesting. Temp tables are local to connections, so these shouldn't stay cached. If memory is still being held, that could be an issue.

**Comment by brianwyka** [comment]
Yes, that's my concern.  The queries basically look like this:

```sql
            CREATE TEMP TABLE t_123456 AS (
                SELECT key, x, y, z
                FROM read_csv('s3://bucket/path/file.csv.gz.')
            );
            COPY t_123456
            TO '/tmp/123456/'
            WITH (FORMAT PARQUET, COMPRESSION GZIP, PARTITION_BY (key), OVERWRITE_OR_IGNORE 1, FILENAME_PATTERN "data.{uuid}");
            DROP TABLE IF EXISTS t_123456;
```

**Comment by brianwyka** [comment]
Just completed another round of testing with removing the long-lived in-memory connection and creating connections fresh every time.  I set the memory limit to `10GB` with `10` threads.  The machine has 10 cores available within Kubernetes.

The memory became exhausted after about 45 minutes of running queries and processing, and then the container restarted due to OOM..
<img width="1039" alt="image" src="https://github.com/duckdb/duckdb/assets/46069296/dbc7f11c-fc9e-481c-95b7-2f9f21a482b1">

**Comment by lnkuiper** [comment]
@brianwyka, thanks for sharing the plot! Some memory seems to be retained at all times, which is definitely a problem

**Comment by brianwyka** [comment]
@lnkuiper , do you need any more information to be able to reproduce or was the above query example enough?

**Comment by lnkuiper** [comment]
@brianwyka I think I will be able to reproduce, but don't have time for this for a while, sadly

**Comment by akoshchiy** [comment]
I think duckdb-java is not bundled with jemalloc, and it could be the reason. https://github.com/duckdb/duckdb/discussions/13075

But also JVM can trim the heap automatically, you can try this settings, which helped me
```
-XX:+UnlockExperimentalVMOptions -XX:TrimNativeHeapInterval=5000
```

**Comment by goenning** [comment]
Have you guys ruled out this being an issue on duckdb instead of the Java driver?

I'm seeing the same behaviour on my end, however I'm using Go instead of Java.
On macOS, after I close the connection, the memory goes up and down as expected.
I have the same code on a Linux Container (with the same dataset) and the memory keeps going up until the container crashes with OOM.

I'll try create a minimal reproducible example.

**Comment by brianwyka** [comment]
@goenning , its possibly due to this: https://github.com/duckdb/duckdb-java/issues/63

**Comment by lnkuiper** [comment]
@brianwyka I've been able to reproduce similar behavior on MacOS with and without jemalloc, but not yet on Linux. I don't think it's related to jemalloc, but you could try with and without to see.

**Comment by louisnow** [comment]
On linux this worked for me, we use k8s and sustained writes -- even after closing the connection and reopening the connection when using the Appender, the memory would not go down.
```
-XX:+UnlockExperimentalVMOptions -XX:TrimNativeHeapInterval=5000
```
Before change
<img width="768" alt="Screenshot 2024-09-04 at 2 32 39 PM" src="https://github.com/user-attachments/assets/ca53873c-fa9b-4146-89f9-c0f0e599f7ac">

The job would die in 30 mins as it's using too much RAM.

After adding the vm options
<img width="791" alt="Screenshot 2024-09-04 at 2 33 41 PM" src="https://github.com/user-attachments/assets/1333ce5f-db6c-4f32-86fc-18a9ca3b3d48">


The memory is actually released. Tested on Duckdb 1.0.0

**Comment by louisnow** [comment]
@lnkuiper Besides calling `malloc_trim` is there any other cleanup that 1.1.0 does? Because right now I close and recreate the connection often to ensure no stale memory is being held by duckdb.

**Comment by lnkuiper** [comment]
DuckDB will keep data cached in memory, depending on your workload. If you are querying DuckDB tables, DuckDB will keep them in memory if possible to speed up subsequent queries. If you are just using temporary data, e.g., reading Parquet files and writing them to CSV, there is nothing to clean up, and there should not be any memory being held. If memory is being held somehow, we would like to see a reproducible issue!

**Comment by louisnow** [comment]
It's a write heavy workload that writes to a duckdb table using the appender class. This table is then finally exported to a hive partitioned parquet table. 

Would this benefit from any changes in 1.1.0?

**Comment by lnkuiper** [comment]
I would expect that dropping the table after you are done with it would free up most cached memory and that resetting the connection is not necessary.

We have improved the performance of our Parquet writer slightly from 1.0.0 to 1.1.0, so hopefully, your workload will improve, but I wouldn't expect the change to be very big.

**Comment by brianwyka** [comment]
@louisnow, does all memory get reclaimed after the workload stops?

**Comment by louisnow** [comment]
@brianwyka It's a short lived pod that spawns up only when there's work to be done and exits immediately after it's done so I can't say for certain. I'll try to test it out maybe and report back.

**Comment by brianwyka** [comment]
Any news or updates on this?  We have our production work loads using CLI API instead of this, so haven't been able to retest this.

**Comment by singh-beant** [comment]
Any updates on this please?

**Comment by lnkuiper** [comment]
@singh-beant We have added jemalloc to the JDBC client, but it somehow won't load. We will try and get this fixed, but most likely not before the v1.3.0 release.

**Comment by louisnow** [comment]
@lnkuiper Is this fixed in v.1.3.0 ? This comment seems to indicate so https://github.com/duckdb/duckdb-java/issues/63#issuecomment-2729201054

**Comment by lnkuiper** [comment]
@louisnow Jemalloc should now be bundled in.

**Comment by brianwyka** [comment]
I did some quick testing last week and I think it has been resolved.  I still need to tune my workload, but from my example, memory was getting reclaimed.
